Assets/Mirror/Runtime/NetworkReader.cs

```diff
@@ -16,8 +16,13 @@ namespace Mirror
     // Note: This class is intended to be extremely pedantic, and
     // throw exceptions whenever stuff is going slightly wrong.
     // The exceptions will be handled in NetworkServer/NetworkClient.
-    public class NetworkReader
+    public class NetworkReader : IDisposable
     {
+        public void Dispose()
+        {
+            NetworkReaderPool.Recycle(this);
```

Comment:
here we introduce dependencies.
hm, that is not ideal.
not everyone will create readers with the pool, and it's not obvious that reader.dispose will do something to the pool.

any idea for alternatives?
maybe NewtorkReaderPooled : NetworkReader?

Reply:
If they don't use the pool they won't likely have `using` either, so it'll be garbage collected.

Reply:
yes, it does increase complexity though.
previously, reader was a standalone thing.
now it depends on a pool. more things can go wrong.

Reply:
pool should be used everywhere.  The using reduces complexity in that we don't have to remember to recycle on every code path.

Reply:
I am not thrilled about the coupling either.

But I think it is worth it,  having a nice "using { ...}"  syntax removes a whole class of bugs,  it works perfectly with exceptions. It ensures that everything that is taken from the pool returns to the pool, and it makes it pretty much impossible to add stuff to the pool by accident.  

So yeah,  the cost is extra coupling,  but the benefit is that all of these problems just can't happen.

Reply:
we need to give this more time. there might be a better solution.
the added coupling isn't good. that's the kind of thing we would've removed when we started forking UNET.
what about the NetworkReaderPooled : NetworkReader suggestion?

Reply:
I worked on NetworkReaderPooled : NetworkReader last night...not working right yet.

Reply:
See #1490